pkg/util/ioutil/ioutil.go

```diff
@@ -7,70 +7,68 @@ package ioutil
 import (
 	"context"
 	"io"
+	"sync/atomic"
 	"time"
 )
 
 // timeoutReader monitors the progress of the io.Reader Read method
-// and calls the cancel function when reading does not progress for
+// and calls the cancelFn function when reading does not progress for
 // the specified time.
 type timeoutReader struct {
-	t *time.Timer
-	r io.Reader
-	n chan int
+	total  atomic.Uint64 // Total number of bytes read until timeout.
+	reader io.Reader
+
+	timer   *time.Timer
+	timeout time.Duration
+
+	done   atomic.Bool
+	cancel context.CancelFunc
 }
 
 // Read implements the io.Reader Read interface.
 func (tr *timeoutReader) Read(p []byte) (int, error) {
-	n, err := tr.r.Read(p)
-
-	v := n
-	if err != nil && tr.t.Stop() {
-		select {
-		case <-tr.t.C:
-		default:
+	n, err := tr.reader.Read(p)
+	if !tr.done.Load() && tr.timer.Reset(tr.timeout) {
+		tr.total.Add(uint64(n))
+		if err != nil {
+			tr.cancel()
+			tr.done.Store(true)
 		}
```

Comment:
Would it make sense to avoid `done` variable? eg:

```
if tr.timer.Reset(tr.timeout) {
		tr.total.Add(uint64(n))
		if err != nil {
			tr.cancel()
			r.timer.Stop()
		}
}
// note: `timer.Stop()` would need to be called in goroutine as well in order to stop timer when context is cancled
```